cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go

```diff
@@ -0,0 +1,678 @@
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider/oci/vendor-internal/github.com/oracle/oci-go-sdk/v55/core"
+)
+
+const (
+	maxAddTaintRetries = 5
+)
+
+var (
+	maxRetryDeadline            time.Duration = 5 * time.Second
+	conflictRetryInterval       time.Duration = 750 * time.Millisecond
+	errInstanceNodePoolNotFound               = errors.New("node pool not found for instance")
+	compiledArmRegexp                         = regexp.MustCompile("\\.A[0-9]+\\.") // Matches node shapes with the pattern '.A<number>.'
+)
+
+// NodePoolManager defines the operations required for an *instance-pool based* autoscaler.
+type NodePoolManager interface {
+	// Refresh triggers refresh of cached resources.
+	Refresh() error
+	// Cleanup cleans up open resources before the cloud provider is destroyed, i.e. go routines etc.
+	Cleanup() error
+
+	// GetNodePools returns list of registered NodePools.
+	GetNodePools() []NodePool
+	// GetNodePoolNodes returns NodePool nodes.
+	GetNodePoolNodes(np NodePool) ([]cloudprovider.Instance, error)
+	// GetNodePoolNodes returns NodePool nodes.
+	GetExistingNodePoolSizeViaCompute(np NodePool) (int, error)
+	// GetNodePoolForInstance returns NodePool to which the given instance belongs.
+	GetNodePoolForInstance(instance ocicommon.OciRef) (NodePool, error)
+	// GetNodePoolTemplateNode returns a template node for NodePool.
+	GetNodePoolTemplateNode(np NodePool) (*apiv1.Node, error)
+	// GetNodePoolSize gets NodePool size.
+	GetNodePoolSize(np NodePool) (int, error)
+	// SetNodePoolSize sets NodePool size.
+	SetNodePoolSize(np NodePool, size int) error
+	// DeleteInstances deletes the given instances. All instances must be controlled by the same NodePool.
+	DeleteInstances(np NodePool, instances []ocicommon.OciRef) error
+	// Invalidate node pool cache and refresh it
+	InvalidateAndRefreshCache() error
+	// Taint with ToBeDeletedByClusterAutoscaler to avoid unexpected CA restarts scheduling pods on a node intended to be deleted before restart
+	TaintToPreventFurtherSchedulingOnRestart(nodes []*apiv1.Node, client kubernetes.Interface) error
+}
+
+type okeClient interface {
+	GetNodePool(context.Context, oke.GetNodePoolRequest) (oke.GetNodePoolResponse, error)
+	UpdateNodePool(context.Context, oke.UpdateNodePoolRequest) (oke.UpdateNodePoolResponse, error)
+	DeleteNode(context.Context, oke.DeleteNodeRequest) (oke.DeleteNodeResponse, error)
+}
+
+// CreateNodePoolManager creates an NodePoolManager that can manage autoscaling node pools
+func CreateNodePoolManager(cloudConfigPath string, discoveryOpts cloudprovider.NodeGroupDiscoveryOptions, kubeClient kubernetes.Interface) (NodePoolManager, error) {
+
+	var err error
+	var configProvider common.ConfigurationProvider
+
+	if os.Getenv(ipconsts.OciUseInstancePrincipalEnvVar) == "true" || os.Getenv(npconsts.OkeUseInstancePrincipalEnvVar) == "true" {
+		klog.Info("using instance principal provider")
+		configProvider, err = auth.InstancePrincipalConfigurationProvider()
```

Comment:
I'm using `InstancePrincipalConfigurationProviderForRegion` in the instance pool implementation and therefore _require_ `OCI_REGION` be set when `OCI_USE_INSTANCE_PRINCIPAL` is set to `true` (see [note in README](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/oci/README.md#environment-variables)).

If both implementations can use `InstancePrincipalConfigurationProvider` we should update the instance pool implementation so the implementations are consistent wrt configuring Instance Principals (and remove the note in the `README`).

Reply:
I see. I was wondering why the OCI_REGION env var was required for instance principals and it's because of the difference in how the configuration provider is initialized. I'll update the instance principal implementation to also use this provider so that OCI_REGION is not needed.

Reply:
Updated the `InstancePrincipalConfigurationProvider` in the Instance Pool implementation to be non region specific